File: frontend/src/api/dataQualityReport.ts
export type DataQualityStatusCounts = {
  PENDING: number;
  HELD: number;
  VALID: number;
  SUSPECT: number;
  EXCLUDED: number;
  REPLACED: number;
};

export interface BoundedDataQualityDetails<T> {
  total: number;
  returned: number;
  has_more: boolean;
  full_set_sha256: string;
  items: T[];
}

export interface DataQualityBulkEntry {
  entry_id: string;
  ordinal: number;
  status: string;
  outcome: string | null;
  status_code: string;
  receipt_id: string | null;
  content_sha256: string | null;
  issue_count: number;
  issues_sha256: string;
  duplicate_of_entry_id: string | null;
  revision_baseline_entry_id: string | null;
}

export interface DataQualityFinalizationEntry {
  bulk_entry_id: string;
  ordinal: number;
  status: string;
  error_code: string | null;
  long_source_file_id: string | null;
  long_ingestion_job_id: string | null;
  long_status: string | null;
  row_version: number;
}

export interface DataQualityResult {
  result_id: string;
  lot_id: string;
  source_file_id: string;
  data_status: string;
  measurement_count: number;
  row_version: number;
  current_data_status_transition_id: string | null;
  current_replacement_transition_id: string | null;
}

export interface DataQualityReplacementLink {
  replacement_id: string;
  predecessor_result_id: string;
  successor_result_id: string;
  predecessor_before_status: string;
  predecessor_after_status: string;
  successor_before_status: string;
  successor_after_status: string;
  decided_at: string;
  candidate_sha256: string;
}

export type DataQualityEvaluationState =
  | "EVALUATED"
  | "NOT_EVALUATED_BY_PHASE"
  | "BLOCKED_BY_INPUT";

export interface InitialDataQualityReport {
  report_version: "initial-data-quality-report-v1";
  report_sha256: string;
  project_key: string;
  batch_id: string;
  supplier_scope: string;
  bulk_status:
    | "STAGED"
    | "PROCESSING"
    | "COMPLETED"
    | "COMPLETED_WITH_EXCEPTIONS"
    | "FAILED";
  inventory: {
    submitted_file_count: number;
    receipt_count: number;
    materialized_source_file_count: number;
    lot_count: number;
    result_count: number;
    measurement_count: number;
  };
  result_status_counts: DataQualityStatusCounts;
  measurement_status_counts: DataQualityStatusCounts;
  bulk_proof: {
    batch_row_version: number;
    terminal: boolean;
    manifest_sha256: string;
    terminal_summary_sha256: string | null;
    outcome_counts: {
      CANDIDATE_READY: number;
      DUPLICATE_CANDIDATE: number;
      MAPPING_REQUIRED: number;
      SCAN_FAILED: number;
      IDENTIFIER_HOLD: number;
      BINDING_HOLD: number;
      VARIATION_REVIEW_REQUIRED: number;
      REVISION_REVIEW_REQUIRED: number;
      ERROR: number;
    };
    unresolved_count: number;
    unresolved_entries_sha256: string;
  };
  finalization_proof: {
    state: "NOT_REQUESTED" | "PRESENT";
    command_id: string | null;
    status: "QUEUED" | "PROCESSING" | "COMPLETED" | "BLOCKED" | null;
    finalization_digest: string | null;
    row_version: number | null;
    total: number;
    pending: number;
    processing: number;
    completed: number;
    blocked: number;
    materialized_job_count: number;
  };
  replacement_proof: {
    transition_count: number;
    predecessor_result_count: number;
    successor_result_count: number;
    transitions_sha256: string;
  };
  details: {
    bulk_entries: BoundedDataQualityDetails<DataQualityBulkEntry>;
    finalization_entries: BoundedDataQualityDetails<DataQualityFinalizationEntry>;
    results: BoundedDataQualityDetails<DataQualityResult>;
    replacement_links: BoundedDataQualityDetails<DataQualityReplacementLink>;
  };
  evaluation_scopes: Array<{
    scope: string;
    state: DataQualityEvaluationState;
    finding_count: number | null;
    evidence_sha256: string | null;
    message: string;
  }>;
  capabilities: {
    read_only: true;
    bounded_details: true;
    official_baseline: false;
    initial_database_gate_complete: false;
    pass_score: false;
    calculations: false;
    thresholds: false;
    ai_used: false;
    scheduler_used: false;
    automatic_state_change: false;
  };
}

export interface DataQualityReportApi {
  getReport(
    batchId: string,
    projectKey: string,
    signal?: AbortSignal,
  ): Promise<InitialDataQualityReport>;
}

const SAFE_FALLBACK_MESSAGE =
  "초기 DB 품질 리포트를 불러오지 못했습니다. 잠시 후 다시 시도해 주세요.";

export class DataQualityReportApiError extends Error {
  readonly code: string;
  readonly statusLabel: string;

  constructor(
    message: string,
    code = "DATA_QUALITY_REPORT_REQUEST_FAILED",
    statusLabel = "초기 DB 품질 리포트 조회 오류",
  ) {
    super(message);
    this.name = "DataQualityReportApiError";
    this.code = code;
    this.statusLabel = statusLabel;
  }
}

export const dataQualityReportApi: DataQualityReportApi = {
  getReport(batchId, projectKey, signal) {
    const query = new URLSearchParams({ project_key: projectKey });
    return requestJson(
      `/api/v1/bulk/batches/${encodeURIComponent(batchId)}/data-quality-report?${query}`,
      { method: "GET", signal },
    );
  },
};

async function requestJson<T>(input: string, init: RequestInit): Promise<T> {
  let response: Response;
  try {
    response = await fetch(input, {
      ...init,
      headers: { Accept: "application/json", ...init.headers },
    });
  } catch (error) {
    if (isAbortError(error)) throw error;
    throw new DataQualityReportApiError(SAFE_FALLBACK_MESSAGE);
  }
  if (!response.ok) throw await safeApiError(response);
  try {
    return (await response.json()) as T;
  } catch {
    throw new DataQualityReportApiError(SAFE_FALLBACK_MESSAGE);
  }
}

async function safeApiError(response: Response): Promise<DataQualityReportApiError> {
  if (response.status < 400 || response.status >= 500) {
    return new DataQualityReportApiError(SAFE_FALLBACK_MESSAGE);
  }
  try {
    const payload: unknown = await response.json();
    if (!isObject(payload) || !isObject(payload.detail)) {
      return new DataQualityReportApiError(SAFE_FALLBACK_MESSAGE);
    }
    const { code, message, status_label: statusLabel } = payload.detail;
    if (
      typeof code !== "string" ||
      typeof message !== "string" ||
      typeof statusLabel !== "string" ||
      !code ||
      !message ||
      !statusLabel
    ) {
      return new DataQualityReportApiError(SAFE_FALLBACK_MESSAGE);
    }
    return new DataQualityReportApiError(message, code, statusLabel);
  } catch {
    return new DataQualityReportApiError(SAFE_FALLBACK_MESSAGE);
  }
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isAbortError(error: unknown): boolean {
  return error instanceof DOMException && error.name === "AbortError";
}

File: backend/app/domain/data_quality_report.py
"""Read-only Phase 2 initial database data-quality report contracts."""

from __future__ import annotations

import re
from dataclasses import dataclass
from enum import StrEnum

_SHA256 = re.compile(r"[0-9a-f]{64}\Z")


class EvaluationState(StrEnum):
    EVALUATED = "EVALUATED"
    NOT_EVALUATED_BY_PHASE = "NOT_EVALUATED_BY_PHASE"
    BLOCKED_BY_INPUT = "BLOCKED_BY_INPUT"


class FinalizationProofState(StrEnum):
    NOT_REQUESTED = "NOT_REQUESTED"
    PRESENT = "PRESENT"


@dataclass(frozen=True, slots=True)
class DataQualityInventory:
    submitted_file_count: int
    receipt_count: int
    materialized_source_file_count: int
    lot_count: int
    result_count: int
    measurement_count: int

    def __post_init__(self) -> None:
        values = (
            self.submitted_file_count,
            self.receipt_count,
            self.materialized_source_file_count,
            self.lot_count,
            self.result_count,
            self.measurement_count,
        )
        if any(value < 0 for value in values):
            raise ValueError("inventory counts cannot be negative")
        if self.receipt_count > self.submitted_file_count:
            raise ValueError("receipt count cannot exceed submitted files")
        if self.materialized_source_file_count > self.receipt_count:
            raise ValueError("materialized files cannot exceed receipts")


@dataclass(frozen=True, slots=True)
class DataStatusCounts:
    pending: int
    held: int
    valid: int
    suspect: int
    excluded: int
    replaced: int

    def __post_init__(self) -> None:
        if any(value < 0 for value in self.values()):
            raise ValueError("data-status counts cannot be negative")

    def values(self) -> tuple[int, ...]:
        return (
            self.pending,
            self.held,
            self.valid,
            self.suspect,
            self.excluded,
            self.replaced,
        )


@dataclass(frozen=True, slots=True)
class BulkOutcomeCounts:
    candidate_ready: int
    duplicate_candidate: int
    mapping_required: int
    scan_failed: int
    identifier_hold: int
    binding_hold: int
    variation_review_required: int
    revision_review_required: int
    error: int

    def __post_init__(self) -> None:
        if any(value < 0 for value in self.values()):
            raise ValueError("Bulk outcome counts cannot be negative")

    def values(self) -> tuple[int, ...]:
        return (
            self.candidate_ready,
            self.duplicate_candidate,
            self.mapping_required,
            self.scan_failed,
            self.identifier_hold,
            self.binding_hold,
            self.variation_review_required,
            self.revision_review_required,
            self.error,
        )


@dataclass(frozen=True, slots=True)
class BulkDataQualityProof:
    batch_row_version: int
    terminal: bool
    manifest_sha256: str
    terminal_summary_sha256: str | None
    outcome_counts: BulkOutcomeCounts
    unresolved_count: int
    unresolved_entries_sha256: str

    def __post_init__(self) -> None:
        if self.batch_row_version < 1 or self.unresolved_count < 0:
            raise ValueError("Bulk proof counters are invalid")
        _sha(self.manifest_sha256, "manifest_sha256")
        if self.terminal_summary_sha256 is not None:
            _sha(self.terminal_summary_sha256, "terminal_summary_sha256")
        _sha(self.unresolved_entries_sha256, "unresolved_entries_sha256")


@dataclass(frozen=True, slots=True)
class FinalizationDataQualityProof:
    state: FinalizationProofState
    command_id: str | None
    status: str | None
    finalization_digest: str | None
    row_version: int | None
    total: int
    pending: int
    processing: int
    completed: int
    blocked: int
    materialized_job_count: int

    def __post_init__(self) -> None:
        counts = (
            self.total,
            self.pending,
            self.processing,
            self.completed,
            self.blocked,
            self.materialized_job_count,
        )
        if any(value < 0 for value in counts):
            raise ValueError("finalization proof counts cannot be negative")
        if self.pending + self.processing + self.completed + self.blocked != self.total:
            raise ValueError("finalization partitions must equal total")
        if self.materialized_job_count > self.completed:
            raise ValueError("materialized jobs cannot exceed completed entries")
        nullable = (self.command_id, self.status, self.finalization_digest, self.row_version)
        if self.state == FinalizationProofState.NOT_REQUESTED:
            if any(value is not None for value in nullable) or any(counts):
                raise ValueError("not-requested finalization proof must be empty")
        else:
            if any(value is None for value in nullable):
                raise ValueError("present finalization proof is incomplete")
            _exact(self.command_id, "command_id", 120)
            if self.status not in {"QUEUED", "PROCESSING", "COMPLETED", "BLOCKED"}:
                raise ValueError("finalization status is invalid")
            if self.finalization_digest is None:
                raise ValueError("finalization digest is missing")
            _sha(self.finalization_digest, "finalization_digest")
            if self.row_version is None or self.row_version < 1:
                raise ValueError("finalization row version is invalid")


@dataclass(frozen=True, slots=True)
class ReplacementDataQualityProof:
    transition_count: int
    predecessor_result_count: int
    successor_result_count: int
    transitions_sha256: str

    def __post_init__(self) -> None:
        if (
            min(
                self.transition_count,
                self.predecessor_result_count,
                self.successor_result_count,
            )
            < 0
        ):
            raise ValueError("replacement proof counts cannot be negative")
        if (
            self.predecessor_result_count != self.transition_count
            or self.successor_result_count != self.transition_count
        ):
            raise ValueError("replacement endpoints must equal transition count")
        _sha(self.transitions_sha256, "transitions_sha256")


@dataclass(frozen=True, slots=True)
class BulkEntryDataQualityDetail:
    entry_id: str
    ordinal: int
    status: str
    outcome: str | None
    status_code: str
    receipt_id: str | None
    content_sha256: str | None
    issue_count: int
    issues_sha256: str
    duplicate_of_entry_id: str | None
    revision_baseline_entry_id: str | None


@dataclass(frozen=True, slots=True)
class FinalizationEntryDataQualityDetail:
    bulk_entry_id: str
    ordinal: int
    status: str
    error_code: str | None
    long_source_file_id: str | None
    long_ingestion_job_id: str | None
    long_status: str | None
    row_version: int


@dataclass(frozen=True, slots=True)
class ResultDataQualityDetail:
    result_id: str
    lot_id: str
    source_file_id: str
    data_status: str
    measurement_count: int
    row_version: int
    current_data_status_transition_id: str | None
    current_replacement_transition_id: str | None


@dataclass(frozen=True, slots=True)
class ReplacementLinkDataQualityDetail:
    replacement_id: str
    predecessor_result_id: str
    successor_result_id: str
    predecessor_before_status: str
    predecessor_after_status: str
    successor_before_status: str
    successor_after_status: str
    decided_at: str
    candidate_sha256: str


@dataclass(frozen=True, slots=True)
class BoundedDataQualityDetails[DETAIL]:
    total: int
    returned: int
    has_more: bool
    full_set_sha256: str
    items: tuple[DETAIL, ...]

    def __post_init__(self) -> None:
        if min(self.total, self.returned) < 0 or self.returned > self.total:
            raise ValueError("bounded detail counts are invalid")
        if self.returned != len(self.items):
            raise ValueError("returned detail count disagrees with items")
        if self.has_more != (self.returned < self.total):
            raise ValueError("bounded detail has_more flag is invalid")
        _sha(self.full_set_sha256, "full_set_sha256")


@dataclass(frozen=True, slots=True)
class DataQualityDetails:
    bulk_entries: BoundedDataQualityDetails[BulkEntryDataQualityDetail]
    finalization_entries: BoundedDataQualityDetails[FinalizationEntryDataQualityDetail]
    results: BoundedDataQualityDetails[ResultDataQualityDetail]
    replacement_links: BoundedDataQualityDetails[ReplacementLinkDataQualityDetail]


@dataclass(frozen=True, slots=True)
class DataQualityEvaluationScope:
    scope: str
    state: EvaluationState
    finding_count: int | None
    evidence_sha256: str | None
    message: str

    def __post_init__(self) -> None:
        _exact(self.scope, "scope", 120)
        _exact(self.message, "message", 1000)
        if self.state == EvaluationState.EVALUATED:
            if self.finding_count is None or self.finding_count < 0:
                raise ValueError("evaluated scope requires a nonnegative finding count")
            if self.evidence_sha256 is None:
                raise ValueError("evaluated scope requires evidence")
            _sha(self.evidence_sha256, "evidence_sha256")
        elif self.finding_count is not None or self.evidence_sha256 is not None:
            raise ValueError("unevaluated scope cannot imply zero findings or evidence")


@dataclass(frozen=True, slots=True)
class DataQualityCapabilities:
    read_only: bool = True
    bounded_details: bool = True
    official_baseline: bool = False
    initial_database_gate_complete: bool = False
    pass_score: bool = False
    calculations: bool = False
    thresholds: bool = False
    ai_used: bool = False
    scheduler_used: bool = False
    automatic_state_change: bool = False

    def __post_init__(self) -> None:
        if not self.read_only or not self.bounded_details:
            raise ValueError("data-quality report read boundaries cannot be disabled")
        if any(
            (
                self.official_baseline,
                self.initial_database_gate_complete,
                self.pass_score,
                self.calculations,
                self.thresholds,
                self.ai_used,
                self.scheduler_used,
                self.automatic_state_change,
            )
        ):
            raise ValueError("data-quality report cannot expose unimplemented capabilities")


@dataclass(frozen=True, slots=True)
class InitialDataQualityReport:
    report_version: str
    report_sha256: str
    project_key: str
    batch_id: str
    supplier_scope: str
    bulk_status: str
    inventory: DataQualityInventory
    result_status_counts: DataStatusCounts
    measurement_status_counts: DataStatusCounts
    bulk_proof: BulkDataQualityProof
    finalization_proof: FinalizationDataQualityProof
    replacement_proof: ReplacementDataQualityProof
    details: DataQualityDetails
    evaluation_scopes: tuple[DataQualityEvaluationScope, ...]
    capabilities: DataQualityCapabilities = DataQualityCapabilities()

    def __post_init__(self) -> None:
        if self.report_version != "initial-data-quality-report-v1":
            raise ValueError("report version is unsupported")
        _sha(self.report_sha256, "report_sha256")
        _exact(self.project_key, "project_key", 64)
        _exact(self.batch_id, "batch_id", 36)
        _exact(self.supplier_scope, "supplier_scope", 200)
        if self.bulk_status not in {
            "STAGED",
            "PROCESSING",
            "COMPLETED",
            "COMPLETED_WITH_EXCEPTIONS",
            "FAILED",
        }:
            raise ValueError("Bulk status is invalid")
        if sum(self.result_status_counts.values()) != self.inventory.result_count:
            raise ValueError("result status counts must equal inventory")
        if sum(self.measurement_status_counts.values()) != self.inventory.measurement_count:
            raise ValueError("measurement status counts must equal inventory")
        if self.details.bulk_entries.total != self.inventory.submitted_file_count:
            raise ValueError("Bulk detail total must equal submitted files")
        if self.details.results.total != self.inventory.result_count:
            raise ValueError("result detail total must equal inventory")


def _sha(value: str, name: str) -> None:
    if not isinstance(value, str) or _SHA256.fullmatch(value) is None:
        raise ValueError(f"{name} must be a lowercase SHA-256")


def _exact(value: str | None, name: str, limit: int) -> None:
    if not isinstance(value, str) or not value or value != value.strip() or len(value) > limit:
        raise ValueError(f"{name} must be an exact nonblank value")
